### PR TITLE
Test deterministic save and restore

### PR DESCRIPTION
**Pull request by mihan69rus**
Added test deterministic save and restore position of 3 objects, which collide with each other.

### CODE REVIEW

**Comment by nbelakovski** [comment]
PR's should have a title and a description. Please update and please retarget against nbelakovski/determinism_changes

File: src/BulletCollision/CollisionDispatch/btCollisionObject.h [hunk 1]
```diff
@@ -25,8 +25,8 @@ subject to the following restrictions:
 #define DISABLE_DEACTIVATION 4
 #define DISABLE_SIMULATION 5
 
-struct	btBroadphaseProxy;
```
Comment (nbelakovski) [inline src/BulletCollision/CollisionDispatch/btCollisionObject.h:28]: Looks like tabs in this file got automatically converted to spaces. Please undo so that this file is not part of the change set.

File: test/BulletDynamics/test_deterministic_save_and_restore.cpp [hunk 2]
```diff
@@ -0,0 +1,221 @@
+#include "../../Extras/Serialize/BulletFileLoader/bFile.h"
+#include <btBulletDynamicsCommon.h>
+#include "../../src/BulletDynamics/Featherstone/btMultiBodyDynamicsWorld.h"
+#include "../../src/BulletDynamics/Featherstone/btMultiBodyConstraintSolver.h"
+#include "../../Extras/Serialize/BulletWorldImporter/btMultiBodyWorldImporter.h"
+#include "../../Extras/Serialize/BulletWorldImporter/btBulletWorldImporter.h"
+#include "../../Extras/Serialize/BulletWorldImporter/btWorldImporter.h"
+#include "../../src/BulletCollision/CollisionDispatch/btCollisionObject.h"
+#include "../../src/BulletCollision/CollisionDispatch/btCollisionWorld.h"
```
Comment (nbelakovski) [inline test/BulletDynamics/test_deterministic_save_and_restore.cpp]: I believe files under src/ are in the include path, so you shouldn't need to prefix them with ../../src

File: test/BulletDynamics/test_deterministic_save_and_restore.cpp [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+#include "LinearMath/btVector3.h"
+#include "LinearMath/btAlignedObjectArray.h"
+#include "../../src/BulletDynamics/Dynamics/btRigidBody.h"
+
+#include "../../examples/CommonInterfaces/CommonRigidBodyBase.h"
```
Comment (nbelakovski) [inline test/BulletDynamics/test_deterministic_save_and_restore.cpp]: Why is this included?

File: test/BulletDynamics/test_deterministic_save_and_restore.cpp [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+	world->setGravity(btVector3(0, -10, 0));
+
+	return world;
+
+}
+
+GTEST_TEST(BulletDynamics, DeterministicSaveRestore)
+{
+
+	static char filename[] = "test_serialize.mixail";
```
Comment (nbelakovski) [inline test/BulletDynamics/test_deterministic_save_and_restore.cpp]: Just use the .bullet extension, when I made the ending a name in our collaboration session it was just to make a point that the file extension did not matter.

File: test/BulletDynamics/test_deterministic_save_and_restore.cpp [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+#include "../../Extras/Serialize/BulletWorldImporter/btWorldImporter.h"
+#include "../../src/BulletCollision/CollisionDispatch/btCollisionObject.h"
+#include "../../src/BulletCollision/CollisionDispatch/btCollisionWorld.h"
+#include "LinearMath/btVector3.h"
+#include "LinearMath/btAlignedObjectArray.h"
+#include "../../src/BulletDynamics/Dynamics/btRigidBody.h"
+
+#include "../../examples/CommonInterfaces/CommonRigidBodyBase.h"
+#include <gtest/gtest.h>
+#define USE_MOTIONSTATE 0
```
Comment (nbelakovski) [inline test/BulletDynamics/test_deterministic_save_and_restore.cpp]: Why are we setting this #define in this test? It seems like it's not used to test anything, can we remove it?

File: test/BulletDynamics/test_deterministic_save_and_restore.cpp [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+GTEST_TEST(BulletDynamics, DeterministicSaveRestore)
+{
+
+	static char filename[] = "test_serialize.mixail";
+	btRigidBody* box_before[3];
+	btCollisionObject* box_after[3];
+	btMultiBodyDynamicsWorld *initial_world = create_btMultiBodyDynamicsWorld();
+
+	// create ground plane and box shapes
+	// don't worry about gui things
```
Comment (nbelakovski) [inline test/BulletDynamics/test_deterministic_save_and_restore.cpp]: Please remove instructional comments like this one, but I think you should keep comments that help explain the flow of the code.

File: test/BulletDynamics/test_deterministic_save_and_restore.cpp [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+	return world;
+
+}
+
+GTEST_TEST(BulletDynamics, DeterministicSaveRestore)
+{
+
+	static char filename[] = "test_serialize.mixail";
+	btRigidBody* box_before[3];
+	btCollisionObject* box_after[3];
```
Comment (nbelakovski) [inline test/BulletDynamics/test_deterministic_save_and_restore.cpp:51]: Why are before/after of different types?

Also a note, it's usually not a good practice to have numbers in code. They're called "magic numbers" because they have no explanation and therefore presumably have some magical properties. They are usually replaced with #define, like
```
#define NUMBER_OF_BOXES 3
btRigidBody *box_before[NUMBER_OF_BOXES];
```
However in this case I don't think it needs to change, I think it's reasonably clear from the code that there's just 3 boxes. Just wanted to add the note so that you keep it in mind as you're writing code in the future.

File: test/BulletDynamics/test_deterministic_save_and_restore.cpp [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+	// create ground plane and box shapes
+	// don't worry about gui things
+
+	btBoxShape* groundShape = new btBoxShape(btVector3(btScalar(50.), btScalar(50.), btScalar(50.)));
+	{
+		btTransform groundTransform;
+		groundTransform.setIdentity();
+		groundTransform.setOrigin(btVector3(0, -50, 0));
+		btScalar mass(0.);
+		bool isDynamic = (mass != 0.f);
```
Comment (nbelakovski) [inline test/BulletDynamics/test_deterministic_save_and_restore.cpp]: This will always be false, so let's get rid of the bool and the associated if statement?

File: test/BulletDynamics/test_deterministic_save_and_restore.cpp [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+				box_before[i] = new btRigidBody(cInfo);
+			#else
+				box_before[i] = new btRigidBody(mass, 0, colShape, localInertia);
+				box_before[i]->setWorldTransform(startTransform);
+			#endif
+			box_before[i]->setUserIndex(-1);
+			initial_world->addRigidBody(box_before[i]);
+			box_before[i]->setLinearVelocity(btVector3(a[2], 0, a[3]));
+		}
```
Comment (nbelakovski) [inline test/BulletDynamics/test_deterministic_save_and_restore.cpp]: If you're adding 3 boxes, I would just add 3 boxes in a row, or better yet write a function to create a box at a given location and with a particular velocity. This for loop with the weird 'a' variable is rather cumbersome to read.

File: test/BulletDynamics/test_deterministic_save_and_restore.cpp [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+	while (box_before[0]->getCenterOfMassPosition().getY() == box_before[1]->getCenterOfMassPosition().getY() &&
+			box_before[1]->getCenterOfMassPosition().getY() == box_before[2]->getCenterOfMassPosition().getY())
+	{
+		time++;
+		initial_world->stepSimulation(delta_t);
+	}
+
+	// serialize
+
+	printf("Serialize in %d time\n", time);
```
Comment (nbelakovski) [inline test/BulletDynamics/test_deterministic_save_and_restore.cpp]: Remove print statements

File: test/BulletDynamics/test_deterministic_save_and_restore.cpp [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+	// step simulation until collision
+
+	const double delta_t = 0.01;
+	int time = 0;
+	while (box_before[0]->getCenterOfMassPosition().getY() == box_before[1]->getCenterOfMassPosition().getY() &&
+			box_before[1]->getCenterOfMassPosition().getY() == box_before[2]->getCenterOfMassPosition().getY())
+	{
+		time++;
+		initial_world->stepSimulation(delta_t);
+	}
```
Comment (nbelakovski) [inline test/BulletDynamics/test_deterministic_save_and_restore.cpp:92]: I see, you're using the fact that the blocks start in the same y position and all start with 0 y-velocity to determine collision. Since what we're actually interested in is whether or not there are any collision manifolds active, let's check for that. You can do this by checking with the dispatcher, i.e. initial_world->getDispatcher()->getNumManifolds(). We should make sure that there are at least 2 collision manifolds present for multiple time steps, because part of the non-determinism comes from collision manifolds that add points to the contact manifold over multiple time steps.

File: test/BulletDynamics/test_deterministic_save_and_restore.cpp [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+
+	printf("Serialize in %d time\n", time);
+	btSerializer* s = new btDefaultSerializer;
+	initial_world->serialize(s);
+	FILE* file = fopen(filename, "wb");
+	fwrite(s->getBufferPointer(), s->getCurrentBufferSize(), 1, file);
+	fclose(file);
+	const btVector3 before_serialize1 = box_before[0]->getCenterOfMassPosition();
+	const btVector3 before_serialize2 = box_before[1]->getCenterOfMassPosition();
+	const btVector3 before_serialize3 = box_before[2]->getCenterOfMassPosition();
```
Comment (nbelakovski) [inline test/BulletDynamics/test_deterministic_save_and_restore.cpp]: Good that you made these const. Let's move these 3 lines to just before the serialize, because we want to be confident that nothing in the serialization pipeline changes the position of the objects.

File: test/BulletDynamics/test_deterministic_save_and_restore.cpp [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+
+	// step simulation until boxes are no longer moving
+
+	for (int i = 0; i < 500; ++i)
+	{
+		time++;
+		initial_world->stepSimulation(delta_t);
+	}
```
Comment (nbelakovski) [inline test/BulletDynamics/test_deterministic_save_and_restore.cpp:117]: We should probably check the position of boxes and make sure they haven't moved for, I dunno, 10(?) frames instead of just looping for a long time.

File: test/BulletDynamics/test_deterministic_save_and_restore.cpp [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+
+	// save position of all boxes
+
+	printf("Save position in %d time\n",time);
+	const btVector3 before_deserialize1 = box_before[0]->getCenterOfMassPosition();
+	const btVector3 before_deserialize2 = box_before[1]->getCenterOfMassPosition();
+	const btVector3 before_deserialize3 = box_before[2]->getCenterOfMassPosition();
```
Comment (nbelakovski) [inline test/BulletDynamics/test_deterministic_save_and_restore.cpp]: I would call these steady_state[123], because what's important here are that the boxes are at rest.

File: test/BulletDynamics/test_deterministic_save_and_restore.cpp [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+	// deserialize
+
+	printf("Deserialize in %d time\n",time);
+	btMultiBodyDynamicsWorld *deserialized_world = create_btMultiBodyDynamicsWorld();
+	btMultiBodyWorldImporter importer(deserialized_world);
+	importer.loadFile(filename);
+	box_after[0] = importer.getRigidBodyByIndex(1);
+	box_after[1] = importer.getRigidBodyByIndex(2);
+	box_after[2] = importer.getRigidBodyByIndex(3);
+	// importer.convertAllObjects(...);
```
Comment (nbelakovski) [inline test/BulletDynamics/test_deterministic_save_and_restore.cpp]: remove

File: test/BulletDynamics/test_deterministic_save_and_restore.cpp [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+	EXPECT_FLOAT_EQ(box_after[0]->getCenterOfMassPosition().getX(), before_serialize1.getX());
+	EXPECT_FLOAT_EQ(box_after[1]->getCenterOfMassPosition().getX(), before_serialize2.getX());
+	EXPECT_FLOAT_EQ(box_after[2]->getCenterOfMassPosition().getX(), before_serialize3.getX());
+	EXPECT_FLOAT_EQ(box_after[0]->getCenterOfMassPosition().getY(), before_serialize1.getY());
+	EXPECT_FLOAT_EQ(box_after[1]->getCenterOfMassPosition().getY(), before_serialize2.getY());
+	EXPECT_FLOAT_EQ(box_after[2]->getCenterOfMassPosition().getY(), before_serialize3.getY());
+	EXPECT_FLOAT_EQ(box_after[0]->getCenterOfMassPosition().getZ(), before_serialize1.getZ());
+	EXPECT_FLOAT_EQ(box_after[1]->getCenterOfMassPosition().getZ(), before_serialize2.getZ());
+	EXPECT_FLOAT_EQ(box_after[2]->getCenterOfMassPosition().getZ(), before_serialize3.getZ());
```
Comment (nbelakovski) [inline test/BulletDynamics/test_deterministic_save_and_restore.cpp]: Why wouldn't we expect these two be equal?

File: test/BulletDynamics/test_deterministic_save_and_restore.cpp [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+/*	EXPECT_FLOAT_EQ(before_deserialize1.getX(),after_deserialize1.getX());
+	EXPECT_FLOAT_EQ(before_deserialize1.getY(),after_deserialize1.getY());
+	EXPECT_FLOAT_EQ(before_deserialize1.getZ(),after_deserialize1.getZ());
+	EXPECT_FLOAT_EQ(before_deserialize2.getX(),after_deserialize2.getX());
+	EXPECT_FLOAT_EQ(before_deserialize2.getY(),after_deserialize2.getY());
+	EXPECT_FLOAT_EQ(before_deserialize2.getZ(),after_deserialize2.getZ());
+	EXPECT_FLOAT_EQ(before_deserialize3.getX(),after_deserialize3.getX());
+	EXPECT_FLOAT_EQ(before_deserialize3.getY(),after_deserialize3.getY());
+	EXPECT_FLOAT_EQ(before_deserialize3.getZ(),after_deserialize3.getZ());
+*/
```
Comment (nbelakovski) [inline test/BulletDynamics/test_deterministic_save_and_restore.cpp]: These seem like the checks that we want right? Why are they commented out?

**Review by nbelakovski — COMMENTED** [review]
Overall a lot of changes needed, and the test doesn't even seem to successfully validate the thing it's testing for. Let's make sure that works ahead of all the other changes, because removing print statements and checking for collision manifolds won't matter if the tests don't pass in the first place.

In order to accomplish this, you may need to rebase onto my branch determinism_changes. There are multiple online guides for rebasing in git, but let me know if you have trouble with that.

File: test/BulletDynamics/test_deterministic_save_and_restore.cpp (hunk 17 above)
Comment (mihan69rus) [reply test/BulletDynamics/test_deterministic_save_and_restore.cpp]: I can't pass all checks at github with failed test

File: test/BulletDynamics/test_deterministic_save_and_restore.cpp [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+	while (box_before[0]->getCenterOfMassPosition().getY() == box_before[1]->getCenterOfMassPosition().getY() &&
+			box_before[1]->getCenterOfMassPosition().getY() == box_before[2]->getCenterOfMassPosition().getY())
+	{
+		time++;
+		initial_world->stepSimulation(delta_t);
+	}
+
+	// serialize
+
+	if (initial_world->getDispatcher()->getNumManifolds() < 2)
```
Comment (nbelakovski) [inline test/BulletDynamics/test_deterministic_save_and_restore.cpp:96]: This should be the condition of the while statement, so that the simulation runs until it is true  (plus a couple more frames after that). Otherwise the test will fail without having actually tested anything...

File: test/BulletDynamics/test_deterministic_save_and_restore.cpp (hunk 17 above)
Comment (nbelakovski) [reply test/BulletDynamics/test_deterministic_save_and_restore.cpp]: Well of course not! That's why I left a comment in the conversations tab about first making sure the tests pass!

Comment (mihan69rus) [reply test/BulletDynamics/test_deterministic_save_and_restore.cpp]: I have some problems with understanding. Test must shows, that boxes after deserialization have position which don't equal position of boxes before serialization, doesn't it?

Comment (nbelakovski) [reply test/BulletDynamics/test_deterministic_save_and_restore.cpp]: No, the test must show that the serialization is deterministic, meaning that with or without serialization, the boxes end up in the same place. Does that make sense?

Comment (mihan69rus) [reply test/BulletDynamics/test_deterministic_save_and_restore.cpp]: So, i think i need some help

Comment (nbelakovski) [reply test/BulletDynamics/test_deterministic_save_and_restore.cpp]: Ok, let's get on Skype and talk about it. Will ping you offline.
